File: common/src/main/java/compasses/expandedstorage/impl/block/misc/CopperBlockHelper.java
package compasses.expandedstorage.impl.block.misc;

import com.google.common.collect.BiMap;
import com.google.common.collect.ImmutableBiMap;
import compasses.expandedstorage.impl.registration.ModBlocks;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockState;

import java.util.Optional;

public class CopperBlockHelper {
    private static final BiMap<Block, Block> OXIDISATION_MAP =
            ImmutableBiMap.<Block, Block>builder()
                          .put(ModBlocks.COPPER_BARREL, ModBlocks.EXPOSED_COPPER_BARREL)
                          .put(ModBlocks.EXPOSED_COPPER_BARREL, ModBlocks.WEATHERED_COPPER_BARREL)
                          .put(ModBlocks.WEATHERED_COPPER_BARREL, ModBlocks.OXIDIZED_COPPER_BARREL)
                          .put(ModBlocks.COPPER_MINI_BARREL, ModBlocks.EXPOSED_COPPER_MINI_BARREL)
                          .put(ModBlocks.EXPOSED_COPPER_MINI_BARREL, ModBlocks.WEATHERED_COPPER_MINI_BARREL)
                          .put(ModBlocks.WEATHERED_COPPER_MINI_BARREL, ModBlocks.OXIDIZED_COPPER_MINI_BARREL)
                          .build();

    private static final BiMap<Block, Block> INVERSE_MAP = OXIDISATION_MAP.inverse();

    private static final BiMap<Block, Block> DEWAXED_MAP =
            ImmutableBiMap.<Block, Block>builder()
                          .put(ModBlocks.WAXED_COPPER_BARREL, ModBlocks.COPPER_BARREL)
                          .put(ModBlocks.WAXED_EXPOSED_COPPER_BARREL, ModBlocks.EXPOSED_COPPER_BARREL)
                          .put(ModBlocks.WAXED_WEATHERED_COPPER_BARREL, ModBlocks.WEATHERED_COPPER_BARREL)
                          .put(ModBlocks.WAXED_OXIDIZED_COPPER_BARREL, ModBlocks.OXIDIZED_COPPER_BARREL)
                          .put(ModBlocks.WAXED_COPPER_MINI_BARREL, ModBlocks.COPPER_MINI_BARREL)
                          .put(ModBlocks.WAXED_EXPOSED_COPPER_MINI_BARREL, ModBlocks.EXPOSED_COPPER_MINI_BARREL)
                          .put(ModBlocks.WAXED_WEATHERED_COPPER_MINI_BARREL, ModBlocks.WEATHERED_COPPER_MINI_BARREL)
                          .put(ModBlocks.WAXED_OXIDIZED_COPPER_MINI_BARREL, ModBlocks.OXIDIZED_COPPER_MINI_BARREL)
                          .build();

    public static Optional<BlockState> getNextOxidisedState(BlockState state) {
        return Optional.ofNullable(OXIDISATION_MAP.getOrDefault(state.getBlock(), null)).map(block -> block.withPropertiesOf(state));
    }

    public static Optional<BlockState> getPreviousOxidisedState(BlockState state) {
        return Optional.ofNullable(INVERSE_MAP.getOrDefault(state.getBlock(), null)).map(block -> block.withPropertiesOf(state));
    }

    public static BiMap<Block, Block> oxidisation() {
        return OXIDISATION_MAP;
    }

    public static BiMap<Block, Block> dewaxing() {
        return DEWAXED_MAP;
    }

    public static Optional<BlockState> getDewaxed(BlockState state) {
        return Optional.ofNullable(DEWAXED_MAP.getOrDefault(state.getBlock(), null)).map(block -> block.withPropertiesOf(state));
    }
}

File: common/src/main/kotlin/semele/quinn/stowage/common/registration/CopperBlockHelper.kt
/*
 * Copyright 2024 Quinn Semele
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package semele.quinn.stowage.common.registration

import com.google.common.collect.BiMap
import com.google.common.collect.ImmutableBiMap
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.state.BlockState
import java.util.Optional

object CopperBlockHelper {
    private var oxidizingMap: BiMap<Block, Block>? = null
    private var oxidizingInverseMap: BiMap<Block, Block>? = null
    private var waxingMap: BiMap<Block, Block>? = null
    private var waxingInverseMap: BiMap<Block, Block>? = null

    fun setOxidizingMap(map: Map<Block, Block>) {
        if (oxidizingMap == null) {
            val immutableMap = ImmutableBiMap.builder<Block, Block>()

            immutableMap.putAll(map)

            oxidizingMap = immutableMap.build()
            oxidizingInverseMap = oxidizingMap!!.inverse()
        }
    }

    fun setWaxingMap(map: Map<Block, Block>) {
        if (waxingMap == null) {
            val immutableMap = ImmutableBiMap.builder<Block, Block>()

            immutableMap.putAll(map)

            waxingMap = immutableMap.build()
            waxingInverseMap = waxingMap!!.inverse()
        }
    }

    fun moreOxidized(state: BlockState): Optional<BlockState> {
        return Optional.ofNullable(oxidizingMap!!.getOrDefault(state.block, null)).map {
            it.withPropertiesOf(state)
        }
    }

    fun lessOxidized(state: BlockState): Optional<BlockState> {
        return Optional.ofNullable(oxidizingInverseMap!!.getOrDefault(state.block, null)).map {
            it.withPropertiesOf(state)
        }
    }

    fun waxed(state: BlockState): Optional<BlockState> {
        return Optional.ofNullable(waxingMap!!.getOrDefault(state.block, null)).map {
            it.withPropertiesOf(state)
        }
    }

    fun unwaxed(state: BlockState): Optional<BlockState> {
        return Optional.ofNullable(waxingInverseMap!!.getOrDefault(state.block, null)).map {
            it.withPropertiesOf(state)
        }
    }

    fun waxingMap(): Map<Block, Block> = waxingMap!!
}